### DISCUSSION TITLE
Enable PHPStan static code analysis on the various repositories

### DISCUSSION THREAD

**Discussion by nhedger** [reactphp/reactphp#469] [Ideas]
### General idea

Enable PHPStan static code analysis on the various ReactPHP repositories.

### Description

I believe the ReactPHP project as a whole and its contributors would benefit from enabling PHPStan. Apart from helping us to catch bugs early, it would also encourage contributors to more strictly type their code with the help of PHPDoc or native type annotations.

Here are some pros:

- Catch bugs early on
- Stricter types
- Better IDE IntelliSense

### Challenges

Since most libraries of the ReactPHP project support a wide range of PHP versions, it might be challenging to set up PHPStan in a way that supports all those versions.

|Project|Version|PHP version range|Supports PHPStan immediately|
|-------|-------|-----------------| --- |
|[async](https://github.com/reactphp/async)|4.x|`>=8.1`| ✅ |
|[cache](https://github.com/reactphp/cache)|1.x|`>=5.3`| ❌ |
|[child-process](https://github.com/reactphp/child-process)|0.6.x|`>=5.3`| ❌ |
|[datagram](https://github.com/reactphp/datagram)|1.x|`>=5.3`| ❌ |
|[dns](https://github.com/reactphp/dns)|1.x|`>=5.3`| ❌ |
|[event-loop](https://github.com/reactphp/event-loop)|1.x|`>=5.3`| ❌ |
|[filesystem](https://github.com/reactphp/filesystem)|0.2.x|`>=7.4`| ✅ |
|[http](https://github.com/reactphp/http)|1.x|`>=5.3`| ❌ |
|[promise](https://github.com/reactphp/promise)|3.x|`>=7.1`| ❌ |
|[promise-stream](https://github.com/reactphp/promise-stream)|1.x|`>=5.3`| ❌ |
|[promise-timer](https://github.com/reactphp/promise-timer)|1.x|`>=5.3`| ❌ |
|[socket](https://github.com/reactphp/socket)|1.x|`>=5.3`| ❌ |
|[stream](https://github.com/reactphp/stream)|1.x|`>=5.3.8`| ❌ |

### Implementation

Given the previous table, it's pretty clear that we would not be able to add `phpstan/phpstan` as a development dependency because the required PHP version would clash.

#### GitHub Actions

PHPStan could be installed with `shivammathur/setup-php@v2` and run only the latest PHP version supported by a given library. Code written for older PHP versions could still be analyzed by specifying the [`phpVersion`](https://phpstan.org/config-reference#phpversion) configuration parameter of PHPStan.

This would add static code analysis to:

- [x] Pushes
- [x] Pull requests 

#### Running locally

I think it's fairly safe to assume that most contributors are running a PHP version superior to `7.1`, so we could pull PHPStan's PHAR into each repository (and exclude it from dist) and run it from there.

---

What are your thoughts on this?

**Comment by WyriHaximus** [comment]
IMHO this would work very well combined with https://github.com/orgs/reactphp/discussions/472 :+1: .

**Reply by nhedger** [reply]
My thoughts exactly!

**Comment by clue** [comment]
@nhedger Excellent initiative, would love to have @phpstan support in @reactphp! I have to admit I also like this would bring us full circle, considering @phpstan also uses @reactphp 🎉

Let's start small: Adding this to GitHub actions only and then only running this on one of the latest PHP versions would probably be a good starting point. I think that unlike PHPUnit, there's little value in also running this against the full test matrix and this would only complicate things.

Adding this to `composer.json` would be a nice bonus, but you're right that we would have to take care of legacy PHP installations then. We may remove this before running `composer install` like https://github.com/clue/reactphp-ssh-proxy/blob/5861005722b9bff9102263aa0eb4e7113c33a29c/.github/workflows/ci.yml#L32 and/or provide additional instructions in our README, but I agree we shouldn't be too concerned about legacy setups anyway (see also #472).

This topic seems to be in good hands here, so I'll leave this up to you to decide. Looking forward to your PR! :shipit:
